src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala

```diff
@@ -0,0 +1,343 @@
+  // should implement HasWeight and HasBias traits.
+  def dWeight():String = throw new DMLRuntimeException("dWeight is not implemented in super class")
+  def dBias():String = throw new DMLRuntimeException("dBias is not implemented in super class")
+  def weight():String = null;
+  def bias():String = null;
+  def shouldUpdateWeight():Boolean = if(weight != null) true else false
+  def shouldUpdateBias():Boolean = if(bias != null) true else false
+  // --------------------------------------------------------------------------------------
+  // Helper methods to simplify the code of subclasses
+  def invokeInit(dmlScript:StringBuilder, returnVariables:List[String], arguments:String*):Unit = {
+    invoke(dmlScript, sourceFileName + "::", returnVariables, "init", arguments.toList)
+  }
+  def invokeForward(dmlScript:StringBuilder, returnVariables:List[String], arguments:String*):Unit = {
+    invoke(dmlScript, sourceFileName + "::", returnVariables, "forward", arguments.toList)
+  }
+  def invokeBackward(dmlScript:StringBuilder, outSuffix:String, resultVariables:List[String],  arguments:String*):Unit = {
+    invoke(dmlScript, sourceFileName + "::", resultVariables.map(_ + outSuffix), "backward", arguments.toList)
+  }
+  // --------------------------------------------------------------------------------------
+}
+
+
+trait IsLossLayer extends CaffeLayer {
+  def computeLoss(dmlScript:StringBuilder, numTabs:Int):Unit
+}
+
+trait HasWeight extends CaffeLayer {
+  override def weight = "W" + id
+  override def dWeight = "dW" + id
+}
+
+trait HasBias extends CaffeLayer {
+  override def bias = "b" + id
+  override def dBias = "db" + id
+}
+
+class Data(val param:LayerParameter, val id:Int, val net:CaffeNetwork, val numChannels:String, val height:String, val width:String) extends CaffeLayer {
+  // -------------------------------------------------
+  override def sourceFileName = null
+  override def init(dmlScript:StringBuilder) = {
+    if(param.hasTransformParam && param.getTransformParam.hasScale) {
+      dmlScript.append("X_full = X_full * " + param.getTransformParam.getScale + "\n")
+    }
+    dmlScript.append("BATCH_SIZE = " + param.getDataParam.getBatchSize + "\n")
+  }
+  var dataOutputShape = ("$num_channels", "$height", "$width")
+  override def forward(dmlScript:StringBuilder, isPrediction:Boolean) = { }
+  override def out = "Xb"
+  override def backward(dmlScript:StringBuilder, outSuffix:String) = { }
+  override def outputShape = (numChannels, height, width)
+  // -------------------------------------------------
+}
+
+
+// ------------------------------------------------------------------
+// weight is ema_mean and bias is ema_var
+// Fuse 
+class BatchNorm(val param:LayerParameter, val id:Int, val net:CaffeNetwork) extends CaffeLayer with HasWeight with HasBias {
+  // val scale =  
+  override def sourceFileName = "spatial_batch_norm"
```

Comment:
This has been renamed: `spatial_batch_norm` --> `batch_norm2d`.  Same with `batch_norm` --> `batch_norm1d`.